### ISSUE TITLE
CS0027 claims `this` cannot be used in an attribute `nameof`, but it compiles just fine

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#82251]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/CS0027-claims-this-cannot-be-used-in-a/11020113)._

---
\[severity:It bothers me. A fix would be nice\]  
Given a method like this:

```  
[MemberNotNull(nameof(this.Name))]
public MyShape SetName(string name)
{
	this.Name = name;
	return this;
}
  
```

CS0027 will (sometimes) be emitted as an **error** in VS2026, with the location referring to `this.` as it appears in the attribute argument. Yet the compilation succeeds anyway.

CS8618 also is emitted as a false warning in this case.

Removing `this.` resolves both diagnostics:

```  
[MemberNotNull(nameof(Name))]
public MyShape SetName(string name)
{
	this.Name = name;
	return this;
}
  
```

This looks like a bug in the compiler / language service.

Repro steps
-----------

```  
git clone https://github.com/AArnott/Nerdbank.MessagePack.git
cd Nerdbank.MessagePack
git checkout b1e27e7a3e5954056e4f2db72d2fc1d8ce1f2053
 .\Nerdbank.MessagePack.slnx
  
```

Look at the error list. You may have to open `MyShape.cs` in order to see it populate with these two false diagnostics.

---
#### Feedback Bot on 12/22/2025, 03:14 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Feedback Bot on 1/30/2026, 09:10 PM: 

<p>This issue is currently being investigated. Our team will get back to you if either more information is needed, a workaround is available, or the issue is resolved.</p>

**Comment by jjonescz** [comment]
This is a regression from https://github.com/dotnet/roslyn/pull/81628#discussion_r2606034681 (cc @AlekseyTs).

**Comment by jjonescz** [comment]
This is actually by design (although previously the compiler allowed that, it was a bug). The reason you are seeing the error in the IDE only is likely because your command-line compiler is older than your IDE compiler.

[nameof spec](https://github.com/dotnet/csharpstandard/blob/5d2c2a2bbc1e93dabc402c033757dd2125b49088/standard/expressions.md#12823-the-nameof-operator) says that some errors from lookup are ignored, but [`this` access](https://github.com/dotnet/csharpstandard/blob/5d2c2a2bbc1e93dabc402c033757dd2125b49088/standard/expressions.md#12814-this-access) is not part of it (`this` is not a name reference, it's a special keyword which is allowed only in very specific contexts, but not in attributes per the language standard).

The fix is simply to remove `this.`:

```diff
- [MemberNotNull(nameof(this.Name))]
+ [MemberNotNull(nameof(Name))]
```

**Comment by CyrusNajmabadi** [comment]
This may be a significant regression.  IIRC the IDE put in extra effort to support 'this' in an attribute precisely because this was allowed, and users were writing it.  If this was longstanding support, it may be good to bring to ldm to decide what to do here:

<img width="1355" height="173" alt="Image" src="https://github.com/user-attachments/assets/03b4d11a-74fe-405e-89d9-fec3ce56c300" />

Comment is slightly wrong, since it is the compiler supporting it.  But it's still something that used to work, that we def know people took advantage of.  So we should at least make sure the break is acceptable.

**Comment by MulleDK19** [comment]
Yay, a **BREAKING** change, with no news about it, breaking every project...

This *should* be allowed. I like to use `this` to make it clear I'm referring to instance members instead of static members. Now I can no longer see that in attributes..

**Comment by CyrusNajmabadi** [comment]
@jjonescz i'm not sure why this was closed.  At the very least, i think the breaking change needs to be assessed and documented :)

It really is breaking people.  And we definitely know people use this feature as they have explicitly asked in the past to make our features work better in the presence of this previously allowed code :)

@jcouv  for breaking change documentation process.

**Comment by jjonescz** [comment]
> i think the breaking change needs to be assessed and documented

That's true, let me reopen for that.

> It really is breaking people.

I think we are past the point of reverting this break for .NET 10.0.200 anyway, but we will discuss this, thanks.

**Comment by rickbrew** [comment]
Yeah this is destroying my codebase. I would definitely like to see this fixed. I have to rollback to the previous version until this is fixed.

**Comment by aromaa** [comment]
Also being effected by this.

Does this also not introduce ambiguity?
```csharp
private string? value;

[MemberNotNull(nameof(value))]
private void Foo(string value) => this.value = value;
```

**Comment by rickbrew** [comment]
> Does this also not introduce ambiguity?

In this particular case does it matter? You get `"value"` either way ¯\\_(ツ)_/¯

**Comment by aromaa** [comment]
> In this particular case does it matter? You get "value" either way ¯\(ツ)/¯

From code perspective, no. But the IDE does care, it thinks you are referring to the parameter instead of the member. Features like "Show All References" no longer display it. What happens if you rename the field or the parameter?

**Comment by rickbrew** [comment]
> From code perspective, no. But the IDE does care, it thinks you are referring to the parameter instead of the member. Features like "Show All References" no longer display it. What happens if you rename the field or the parameter?

Oh, good points, I did not think of that

**Comment by hamarb123** [comment]
> Does this also not introduce ambiguity?
> 
> private string? value;
> 
> [MemberNotNull(nameof(value))]
> private void Foo(string value) => this.value = value;

There is a cheating way to work around this:
```csharp
using System;
using System.Diagnostics.CodeAnalysis;
public class C
{
    private string? value;
    private static C _staticC;

    [MemberNotNull(nameof(_staticC.value))]
    private void Foo(string value) => this.value = value;
}
```
(the other form refers to the parameter).

But this seems unideal as a solution lol - maybe there's a variation that works without adding junk to the assembly & making it nasty, but not certain - however if you have some other instance of the type in scope (e.g., parameter, field on anything you can access, etc.) then you can do it on that instance.

**Comment by CyrusNajmabadi** [comment]
> What happens if you rename the field or the parameter?

Note: nameof confirms a specific name.  But doesn't always guarantee that this will work.  A trivial example since the start of the feature was simply:

```c#
class C
{
    void M() { }

    void M(int i) { }

    void G()
    {
        Console.WriteLine(nameof(M));
    }
}
```

Ambiguity in nameof names is part and parcel of things.  Furthermore, it's never been a given that all symbols are capable of being referenced in nameof.  For example:

```c#
void F(int p)
{
    void G(int p)
    {
        // Simply no way at all to reference 'p' from F's signature.
    }
}
```

**Comment by ChristophHornung** [comment]
Why was that breaking change even made?

I see no upside in disallowing `this` while on the downside I now have to go through dozens of projects and either try to stay on an older .NET version or fix a long list of compiler errors.
Especially in attributes `[MemberNotNull(nameof(this.serviceProvider))]` and similar attributes this was a common pattern.

**Comment by CyrusNajmabadi** [comment]
> Why was that breaking change even made?

1. The compiler was out of spec with the language itself. 
2. Supporting this was possible due to buggy code in the compiler related to attributes. Maintaining those bugs while trying to properly implement other features was costly and difficult. 
3. Working around the break is easy.

**Comment by Takym** [comment]
I've opened a discussion at the dotnet/csharplang repo.
https://github.com/dotnet/csharplang/discussions/10059

**Comment by jnm2** [comment]
The "wrong thing gets renamed" issue seems quite real to me. Nameof can't disambiguate every situation, but there are specific reasons to want it here.

**Comment by CyrusNajmabadi** [comment]
> Nameof can't disambiguate every situation

I think this is the critical thing for me.  It's *always* been something helpful in the language, but not suitable for all situations.  There are already many named things that nameof may not be able to reference depending on context (including hiding, or unreachable symbols).   I don't want strange compiler bugs to be used to workaround that.  We should instead decide *if* those scenarios are important for nameof to work in and decide what the right lang rules are to make those possible.

**Comment by Takym** [comment]
I have already opened the discussion about the language specification at <https://github.com/dotnet/csharplang/discussions/10059>. Also, my opinion is not that "strange compiler bugs to be used".
